### DISCUSSION TITLE
Percentage of bin in bar chart

### DISCUSSION THREAD

**Discussion by LSchueler** [vega/altair#2479] [Q&A] [answered]
Hi and a big thank you for altair!

I have a table which I want to visualize as a bar chart. On the X axis, I want to bin the values into the given intervals (they are strings). On the Y axis, I want to give the percentage of positive values with respect to the individual bin. As an example for the example table below, for the `51 - 100` bin I want to show `50%`, as there are 2 out of 4 positive values. The examples I found always calculate the relative values with respect to the total number of values.
I am unsure what transform methods to use and how to use them to get the wanted results.
It would be really great if someone could give me a hint!

```
ind  value                                        bins
0    Negative                                    0 - 50
1    Negative                                    0 - 50
2    Negative                                    0 - 50
3    Negative                                    0 - 50
4    Negative                                    0 - 50
5    Negative                                    0 - 50
6    Positive                                  51 - 100
7    Positive                                  51 - 100
8    Negative                                  51 - 100
9    Negative                                  51 - 100
10   Positive                                    0 - 50
11   Positive                                    0 - 50
12   Positive                                    0 - 50
13   Positive                                    0 - 50
14   Positive                                    0 - 50
15   Positive                                    0 - 50

```

**Comment by joelostblom** [comment]
There is `transform_aggregate`, [but there is not proportion/percentage operation](https://vega.github.io/vega-lite/docs/aggregate.html#ops). You might be able to put something together using `transform_calculate`, but I am not sure how on the top of my head. You could also use pandas like this: `df.groupby('bins')['value'].value_counts(normalize=True).xs('Positive', level='value').reset_index()`.

**Comment by jakevdp** [comment, marked as answer]
You can do this by roughly following the [Percent of Total](https://altair-viz.github.io/gallery/percentage_of_total.html) example, along with some pre-processing of your data in a calculate transform. For example:
```python
import pandas as pd
import io
import altair as alt

data = """\
ind  value                                        bins
0    Negative                                    0 - 50
1    Negative                                    0 - 50
2    Negative                                    0 - 50
3    Negative                                    0 - 50
4    Negative                                    0 - 50
5    Negative                                    0 - 50
6    Positive                                  51 - 100
7    Positive                                  51 - 100
8    Negative                                  51 - 100
9    Negative                                  51 - 100
10   Positive                                    0 - 50
11   Positive                                    0 - 50
12   Positive                                    0 - 50
13   Positive                                    0 - 50
14   Positive                                    0 - 50
15   Positive                                    0 - 50
"""
source = pd.read_csv(io.StringIO(data), delimiter='\s{2,}', engine='python')

alt.Chart(source).transform_calculate(
    is_positive=(alt.datum.value == "Positive")
).transform_joinaggregate(
    total_pos='sum(is_positive)',
    total='count()',
    groupby=['bins']
).transform_calculate(
    percent=alt.datum.total_pos / alt.datum.total
).mark_bar().encode(
    y='bins:N',
    x=alt.X('percent:Q', axis=alt.Axis(format='%'))
)
```
![visualization - 2021-07-30T053518 315](https://user-images.githubusercontent.com/781659/127653724-73d8d2a2-264a-4094-9e38-3892fe252ccb.png)

**Reply by LSchueler** [reply]
Thank you so much for helping me out!
